File: leetcode/2389/longest_subsequence_limited_sum.cpp
// 6ms | 16.6mb
// 96.15% | 16.84%
#include <vector>
#include <iostream>

using namespace std;

class Solution {

public:
  vector<int> answerQueries(vector<int> &nums, vector<int> &queries) {
    sort(nums.begin(), nums.end());

    // Calculate running sum
    vector<int> running_sum;
    int rs = 0;
    for (int num : nums) {
      rs += num;
      running_sum.push_back(rs);
    }

    vector<int> result;
    for (int query : queries) {
      int result_idx = binarySearch(running_sum, query);
      result.push_back(result_idx + 1);
    }

    return result;
  }

  int binarySearch(vector<int> &array, int value) {
    int start = 0;
    int end = array.size() - 1;

    int mid = 0;
    while (start <= end) {

      mid = (start + end) / 2;

      if (array[mid] == value) {
        return mid;
      } else if (array[mid] < value) {
        start = mid + 1;
      } else {
        end = mid - 1;
      }
    }

    if (array[mid] > value) {
      mid -= 1;
    }
    return mid;
  }
};

File: leetcode/2389/longest_subsequence_limited_sum.py
# 95ms | 16.86mb
# 60.09% | 87.76%
class Solution:
    def answerQueries(self, nums: list[int], queries: list[int]) -> list[int]:
        result = []
        nums.sort()

        self.running_sum = []
        rs = 0
        for num in nums:
            rs += num
            self.running_sum.append(rs)

        for query in queries:
            rs_idx = self.binarySearch(query)
            result.append(rs_idx + 1)
        return result

    def binarySearch(self, q):
        start = 0
        end = len(self.running_sum) - 1
        while start <= end:
            mid = (start + end) // 2
            if self.running_sum[mid] == q:
                return mid
            elif self.running_sum[mid] < q:
                start = mid+1
            else:
                end = mid - 1

        if self.running_sum[mid] > q:
            return mid - 1
        return mid
